### ISSUE TITLE
Creating a contact (example request)

### ISSUE THREAD

**Issue by maantje** [Garethp/php-ews#20]
Can you create an example on how to create a contact? Currently I'm trying the following, but it is returning this error:
`
The request failed schema validation: The element 'Items' in namespace 'http://schemas.microsoft.com/exchange/services/2006/messages' has incomplete content. List of possible elements expected: 'Item, Message, CalendarItem, Contact, DistributionList, MeetingMessage, MeetingRequest, MeetingResponse, MeetingCancellation, Task, ReplyToItem, ForwardItem, ReplyAllToItem, AcceptItem, TentativelyAcceptItem, DeclineItem, CancelCalendarItem, RemoveItem, SuppressReadReceipt' in namespace 'http://schemas.microsoft.com/exchange/services/2006/types'.`

Code:

>         $ews = API::withUsernameAndPassword (
>             $this->server_address,
>             $this->server_username,
>             $this->server_password,
>             [
>                 'impersonation' => Auth::user()->employee->email,
>                 'version' => "Exchange2007"
>             ]
>         );
>     
>         $request = new EWSType_CreateItemType();
> 
>         $contact = new EWSType_ContactItemType();
>         $contact->Initials = $this->relation->initials;
>         $contact->GivenName = $this->relation->first_name;
>         $contact->MiddleName = $this->relation->insertion;
>         $contact->Surname = $this->relation->last_name;
>         $contact->CompleteName = User::findRelationFullName($this->relation);
> 
>         $emails = [];
> 
>         // relation email address
>         if (!empty($this->relation->email)) {
>             // create an email address
>             $email = new EWSType_EmailAddressDictionaryEntryType();
>             $email->Key = new EWSType_EmailAddressKeyType();
>             $email->Key->_ = EWSType_EmailAddressKeyType::EMAIL_ADDRESS_1;
>             $email->_ = $this->relation->email;
>             $emails[] = $email;
>         }
> 
>         // organization email address
>         if (!empty($this->relation->organization->email)) {
>             // create an email address
>             $email = new EWSType_EmailAddressDictionaryEntryType();
>             $email->Key = new EWSType_EmailAddressKeyType();
>             $email->Key->_ = EWSType_EmailAddressKeyType::EMAIL_ADDRESS_2;
>             $email->_ = $this->relation->organization->email;
>             $emails[] = $email;
>         }
> 
>         if (!empty($emails)) {
>             // set the email
>             $contact->EmailAddresses = new EWSType_EmailAddressDictionaryType();
>             $contact->EmailAddresses->Entry = $emails;
>         }
> 
>         // create an address
>         $address = new EWSType_PhysicalAddressDictionaryEntryType();
>         $address->Key = new EWSType_PhysicalAddressKeyType();
>         $address->Key->_ = EWSType_PhysicalAddressKeyType::HOME;
>         $address->Street = $this->relation->organization->addresses->first()->street;
>         $address->City = $this->relation->organization->addresses->first()->city;
>         //$address->State = 'PA';
>         $address->PostalCode = $this->relation->organization->addresses->first()->zipcode;
>         $address->CountryOrRegion = $this->relation->organization->addresses->first()->country;
> 
>         // set the address
>         $contact->PhysicalAddresses = new EWSType_PhysicalAddressDictionaryType();
>         $contact->PhysicalAddresses->Entry[] = $address;
> 
>         if (!empty($this->relation->phone_number_direct)) {
>             // create a phone number
>             $phone = new EWSType_PhoneNumberDictionaryEntryType();
>             $phone->Key = new EWSType_PhoneNumberKeyType();
>             $phone->Key->_ = EWSType_PhoneNumberKeyType::HOME_PHONE;
>             $phone->_ = $this->relation->phone_number_direct_code . $this->relation->phone_number_direct;
> 
>             // set the phone number
>             $contact->PhoneNumbers = new EWSType_PhoneNumberDictionaryType();
>             $contact->PhoneNumbers->Entry[] = $phone;
>         }
> 
>         // set the "file as" mapping to "first name last name"
>         $contact->FileAsMapping = new EWSType_FileAsMappingType();
>         $contact->FileAsMapping->_ = EWSType_FileAsMappingType::FIRST_SPACE_LAST;
> 
>         $request->Items->Contact[] = $contact;
> 
>         if (!empty($this->relation->function)) {
>             $contact->JobTitle = $this->relation->function;
>         }
> 
>         $contact->CompanyName = $this->relation->organization->name;
>         $result = $ews->createItems($contact);

**Comment by Garethp** [comment]
Yeah, all that? That's why I wrote this instead. Rather than doing that, we can work with arrays. I haven't gotten around to making creating a contact simple yet, so you'll have to create it manually. [Here's](https://msdn.microsoft.com/en-us/library/office/aa563318(v=exchg.140).aspx) the Microsoft documentation on creating a contact. If you scroll down a bit, you can see the XML needed, so we'll try to match it with an array. It *should* look something like this

```php
$result = $ews->createItems(array (
    'Contact' => array(
        'GivenName' => 'John',
        'Surname' => 'Smith',
        'Initials' => 'J S',
        'CompleteName' => 'John Smith',
        'EmailAddress' => array(
            'Entry' => array ('Key'=>'EmailAddress1', '_value' => 'john.smith@gmail.com')
        )
    )
));
```

And so on. You can pretty much just build your request directly in array's this way. Alternatively, you can create a `Type\ContactItemType` object and pass that in. Like this

```php
$contact = new Type\ContactItemType();

$contact->setGivenName('John')
    ->setSurname('Smith')
    ->setInitials('J S')
    ->addEmailAddress(Type\EmailAddressDictionaryType::buildFromArray(array('key' => 'EmailAddress1', '_value' => 'john.smith@gmail.com')))

$result = $ews->createItems(array('Contact' => $contact));
```

The second should work, but it's also untested. If you go for the second, you can take a look at the file for ContactType to see what you need to pass in for each value.

Just an FYI, I won't be available tomorrow, so if I probably won't be able to answer any more questions till Monday

**Comment by Garethp** [comment]
Oh, and just FYI, because this is probably what's going to come up at some point next and I may as well let you know now. You're going to have to perform updates manually as well. Unlike creates however, you can't just update with an object. **However**, you can build an update out of an array. For example, [here](https://github.com/Garethp/php-ews/blob/master/src/Calendar/CalendarAPI.php#L159) is how I update calendar items. [here](https://github.com/Garethp/php-ews/blob/master/src/API.php#L246) is the function that takes an array of Key=>Value objects and turns them in to the fieldUri things you want. So updating a contact will *probably* look something like this

```php
        $changes = array('GivenName' => 'Jane', 'CompleteName' => 'Jane Smith');

        //Create the request
        $request = array(
            'ItemChange' => array(
                'ItemId' => $itemId->toArray(),
                'Updates' => array(
                    'SetItemField' => $this->buildUpdateItemChanges('ContactItem', 'contact', $changes)
                )
            )
        );
        $items = $this->updateItems($request);
```

Also, take note of the `$itemId->toArray()`, it'll save you from having to do something like `'ItemId' => array('Id' => $itemId->getId(), 'ChangeKey' => $itemId->getChangeKey())`

Also, so you *have* an Id and ChangeKey, and you want to turn that in to an `ItemIdType`, you can do so like this `$itemId = new ItemIdType($id, $changeKey);`

Finally, if you want to see any more examples of how to do things, take a look in Calendar and Mail API's. They interact with the base API in much the same way you'll have to with ContactAPI.

Though I do plan on improving ContactAPI in the coming weeks.

**Comment by baaskoen** [comment]
Thank you for your effort! 

The following code makes it much cleaner.
```
$result = $ews->createItems(array (
    'Contact' => array(
        'GivenName' => 'John',
        'Surname' => 'Smith',
        'Initials' => 'J S',
        'CompleteName' => 'John Smith',
        'EmailAddress' => array(
            'Entry' => array ('Key'=>'EmailAddress1', '_value' => 'john.smith@gmail.com')
        )
    )
));
```

However, I'm still getting this error:
`Set action is invalid for property.`

**Comment by Garethp** [comment]
Just an FYI, my examples were a bit off (You can't set Initials or CompleteName when creating a contact, those are set automatically. Also, updating EmailAddress is a bit of a special case), but I've now released `0.7.5` which makes it a bit easier, I've also added those [examples](https://github.com/Garethp/php-ews/tree/master/examples/contacts) that you were after. If you have any further issues, please feel free to either add a new issue or just respond to this one
